src/cradle/plugins/secondary_cache/http/http_cache.cpp

```diff
@@ -1,68 +1,136 @@
- * the server not to check that blob contents encode an ActionResult, by
- * passing --disable_http_ac_validation=1.
- * An alternative would be to simulate Bazel, and store an ActionResult
- * referring to the blob in the CAS.
+ * Like the memory (immutable) and local disk caches, the HTTP cache implements
+ * a two-phase solution, using two subcaches, called the Action Cache (AC) and
+ * the Content Addressable Storage (CAS), respectively. bazel-remote already
+ * implements these, via the /ac/<key> and /cas/<key> endpoints.
+ *
+ * An Action corresponds to resolving a request. Records in the Action Cache
+ * are indexed by SHA-2 strings that uniquely identify a request. bazel-remote
+ * normally expects the Action Cache to contain ActionResult's; however, in a
+ * CRADLE context, converting digests to and from ActionResult's would be a
+ * major complication without any advantages. The easiest solution is to store
+ * digests under /ac/<key>, and instruct the server not to check that blob
+ * contents encode an ActionResult, by passing --disable_http_ac_validation.
+ *
+ * The CAS stores the result values, indexed by unique digests over those
+ * values. Thus, if two different requests result in the same value, the
+ * corresponding AC records will reference the same CAS record.
+ * A CAS key is the lowercase SHA256 hash of the stored value. This is
+ * identical between CRADLE and Bazel.
+ * A CAS value is a blob that serializes the actual value. Serialization
+ * details are up to the HTTP cache client.
  */
 
+namespace {
+
 std::string
-make_url(int port, std::string const& key)
+make_url(int port, std::string const& cache_name, std::string const& key)
+{
+    return fmt::format("http://localhost:{}/{}/{}", port, cache_name, key);
+}
+
+http_request
+make_cache_get_request(
+    int port, std::string const& cache_name, std::string const& key)
+{
+    return make_get_request(
+        make_url(port, cache_name, key), {{"Accept", "*/*"}});
+}
+
+http_request
+make_ac_get_request(int port, std::string const& ac_key)
 {
-    return fmt::format("http://localhost:{}/ac/{}", port, key);
+    return make_cache_get_request(port, "ac", ac_key);
 }
 
 http_request
-make_cache_get_request(int port, std::string const& key)
+make_cas_get_request(int port, std::string const& digest)
 {
-    return make_get_request(make_url(port, key), {{"Accept", "*/*"}});
+    return make_cache_get_request(port, "cas", digest);
 }
 
 http_request
-make_cache_put_request(int port, std::string const& key, blob value)
+make_cache_put_request(
+    int port,
+    std::string const& cache_name,
+    std::string const& key,
+    blob value)
 {
     return http_request{
         http_request_method::PUT,
-        make_url(port, key),
+        make_url(port, cache_name, key),
         {{"Accept", "*/*"}},
         std::move(value),
         none};
 }
 
+http_request
+make_ac_put_request(
+    int port, std::string const& key, std::string const& digest)
```

Comment:
It seems that `digest` should be passed in by value and moved into the call to `make_blob`.

Reply:
Fixed